### PR TITLE
[9388] add setting that keeps sidebar from toggling folders/opening files

### PR DESCRIPTION
**Pull request by hpcsc**
@bpasero : please help to review

This is to implement the feature request #9388 

**Request**: Provide a setting so that only double click opens a file in the editor or expands a folder

**Implementation**:

- add setting `workbench.sideBar.openOnSelect`, default to `true`, which is existing behaviour of VSCode: toggle folders and preview/open files when selected. When set to `false`, only highlight folders and files when selected and toggle/preview when double click
- not sure whether I need to provide i18n translation for the setting

### CODE REVIEW

**Comment by bpasero** [comment]
@hpcsc your setting is very generic but only works in the Explorer. if a user wants to only open files on double click, wouldn't this setting need to be supported in every tree/list control where something opens on click? For example:
* search results
* git changes
* problems view
* extensions
* etc

**Comment by hpcsc** [comment]
@bpasero you are right, I didn't consider those scenarios yet. That will require a bit more changes. However do you think this setting should be applicable to those places as well, I mean in terms of user experience?

**Comment by bpasero** [comment]
@hpcsc I am not sure, maybe best to ask the people that request this feature?

**Comment by hpcsc** [comment]
@bpasero I search for `TreeDefaults.DefaultController` (where `onLeftClick()` is defined) and there are 11 controllers inheriting from it at the moment:

- `openEditorsViewer`: for Open Editors pane in explorer section
- `explorerViewer`: for explorer pane
- `searchResultsView`: for search result pane
- `changesViewer`: for git changes pane
- `dependenciesViewer`: for extensions tab
- `markersTreeController`: for problems view
- `quickOpenWidget`: I guess this is when I press `Ctrl/Cmd + P`, could you confirm?
- `referencesWidget`: **not sure what's the purpose of this widget**
- `debugHover`: **not sure what's the purpose of this widget**
- `debugViewer`: **not sure what's the purpose of this widget**
- `treeExplorerViewer`: **not sure what's the purpose of this widget**

And beside that, I seem not be able to find controllers for 4 panes in debugger view: variables, watch, call stack, break points. Could you point me to correct places?

I'm just seeing whether it's safe to change `TreeDefaults.DefaultController.onLeftClick()` since that will solve most of the places

**Comment by bpasero** [comment]
* `quickOpenWidget`: yes this is the picker control (e.g. Cmd+P)
* `referencesWidget`: this is when you invoke "Find References" in the editor and see a tree with results embedded
* `debugHover`: this is when you hover over a rich object while debugging a tree is shown that has expandable nodes to show more data
* `debugViewer`: there are multiple viewers within this file for all the various trees of the debug viewlet, for example callstack, variables, breakpoints and watch expressions that all are trees
* `treeExplorerViewer`: this is currently not being used but in the future is intended to be a tree that extensions can use to show a tree

> And beside that, I seem not be able to find controllers for 4 panes in debugger view: variables, watch, call stack, break points. Could you point me to correct places?

Those are all inside `debugViewer`

I agree it would make sense to have most functionality in the base controller, however we talk about a setting here that you get from the `IConfigurationService` and on the layer of the tree there is no such access to the service so it has to be an option that you set for the tree if you want to have it on that layer.

Extensions are not enabled when you run "out of sources". But you can still copy extensions over by:
* from your main Code version open the extensions folder (F1 > Open extensions folder)
* copy some or all
* paste them into the extensions folder of the dev instance

**Comment by hpcsc** [comment]
@bpasero I just pushed some changes:
- rename the setting as discussed
- add `openMode` to `IControllerOptions` so this setting is kept at base class `DefaultController` in `treeDefaults`
- have a `TreeControllerBase` with logic to automatic updating setting when it's changed by user
- change behaviours of explorers, open editors, find reference results to use the above structure

However I realise the way the code handles editor opening when an item is clicked is not consistent at the moment: 
- For explorers, open editors etc, it handles that in controllers. 
- But for some other places, it handles in viewlet. For example search result pane, it listens for tree selection event in `SearchViewlet.createSearchResultsView` and open item there in the `SearchViewlet`. This makes it challenging to use structure I have above.

So I'm thinking of moving the logic in `treeDefaults.DefaultController` and `TreeControllerBase` to a singleton service so it can be injected and used by either controllers or viewlet. What do you think?

Also, is there any further documentation on vscode code structure? E.g. at the moment I'm not clear the structure of viewlet, widget, controller etc and what to read more about that

**Comment by bpasero** [comment]
Give me some time to think about this again, currently I am busy with other things this week.

**Comment by hpcsc** [comment]
@bpasero sure, no worries

**Comment by hpcsc** [comment]
@bpasero Pushed some more changes. Places that I have modified so far:
- Explorer tab - explorer
- Explorer tab - open editors pane
- Debugger tab - Variables pane: **for leaf node, double click still turns the node into text box to edit variable value**
- Debugger tab - Watch pane
- Debugger tab - Call stack pane
- Debugger tab - Breakpoints pane
- Problems view
- Find reference results: **this view already has existing behaviour associated with double click (go to file location) so I still keep that behaviour in the leaf node of the tree**

Pending places:
- Search results: click behaviour is handled in viewlet instead of controller, pending discussion
- Git changes pane: click behaviour is handled in viewlet instead of controller, pending discussion
- Extensions tab
- DebugHover: still trying to find this in the UI
- treeExplorerViewer: **not in use, but for future usage, better to modify this as well?**

**For quick open, I think it makes sense to keep the current behaviour, even in double click mode. Since it's quick open, most people will expect it to be opened in single click in my opinion**

**Comment by bpasero** [comment]
A heads up: I looked through all our users of the tree throughout the code and unfortunately (as already mentioned), not every tree is handling opening in the same way. Specifically:
* some implement `Controller` and do the opening from `onLeftClick`
* some just react on the selection event and open from there

Imho all tree users should convert to one solution that we can then also use to bring in the setting for opening on single or double click.

Since some of the tree API was inspired by Eclipse's tree API, I think another concept of theirs would make sense for us: `IOpenListener` ([example](http://www.programcreek.com/java-api-examples/index.php?class=org.eclipse.jface.viewers.TreeViewer&method=addOpenListener)).

The idea is that the tree (actually not the tree widget but a higher level concept called `StructuredViewer`) itself emits an event when it thinks an element should open. For Eclipse there is actually more cases when this can happen because of this:

image

There are more settings that control if an item should open or not, mainly:
* open on hover (that is a bit weird to be fair)
* open on keyboard navigation (that is interesting!)

So they abstract everything out behind this one listener so that every tree user can deal with this without having to worry about the options. 

Just wanted to throw this in what my current thinking is how to tackle this best.

**Comment by hpcsc** [comment]
@bpasero

Do you mean we have an interface:
```
public interface IOpenListener {
    void open(OpenEvent event);
}
```

and an implementation:
```
public class OpenListener implements IOpenListener {
    constructor(configurationService: IConfigurationService) {
         //register for setting change notification and update key
    }

   public void open(OpenEvent event) {
        const isDoubleClick = event.details == 2;
        if(this.isInSingleOpenMode() || isDoubleClick) {
             //open editor or expand/collapse folder
        }
   }
}
```

And in viewlets/controllers, we do:
```
viewlet.addOpenListener(new OpenListener(configurationService))
```

or
```
controller.addOpenListener(new OpenListener(configurationService));
```

then when it's time to open editor in viewlet or controller, it emits `OpenEvent`

Do I understand it correctly?

I think there are some further complications that I can think of:
- Some trees open current selected file in editor, some other trees open different types of editor (.e.g. git changes compare side by side editor or when clicking on an extension to open extension details). If we do like above, for each type, we will need a different concrete class that implements IOpenListener and viewlet/controller needs to be aware of which concrete class to create during `addOpenListener()`?
- Even if the trees are just opening file editor, the way they open it is also quite different from each other. It might be quite challenging to have a single place that handles all the opening. I might be wrong here

**Comment by bpasero** [comment]
I think downstream clients of the tree should register to a new "open" event and not care about any configuration service and just handle the opening. The tree component has to provide this event based on the configuration of the user that applies. This would be the ideal scenario for me because it allows us to fine tune the event in certain cases where e.g. we actually want to open on keyboard selection change (for example search results, problems view).

**Comment by hpcsc** [comment]
@bpasero Sorry I'm a bit lost. Is my sample code in the previous reply closed to what you are describing? If not, could you elaborate a bit more on your approach? Some sample code will be great

**Comment by hpcsc** [comment]
@bpasero just take some time to think about your suggestion, here is my understanding, correct me if I'm wrong:
- Centralize all the current opening logic into one place (let's say in `DefaultController.onLeftClick()` from `treeDefaults.ts` ). This will include the logic to get configuration setting, update configuration setting if changed, etc. When someone clicks left mouse, this will check whether it's in single click or double click mode and emits a corresponding `OpenEvent`
- `DefaultController` also expose function `addOpenListener()` so anyone interested can listen to that event
- `WorkbenchEditorService` (or other services) can register as open listener for `DefaultController` and will trigger opening of editor when `OpenEvent` is broadcasted

However this can be quite a major change and I don't have complete view of the system to tell whether it's feasible. Especially in centralizing opening logic into one place: there are many customized behaviors in different sub classes of `DefaultController` that I'm not yet sure how to merge it into one place

**Review by bpasero — CHANGES_REQUESTED** [review]

**Comment by bpasero** [comment]
Maybe instead of introducing a new `OPEN` event, we just rely on the existing `SELECTION` event and adopt those tree users that currently do not open on `SELECT` to actually do that.

The idea in the tree was always that a selected item is also opened. The controller could check for the open mode and only set the selection based on double click or not. 

We already use this model in some places (e.g. search, git).

**Comment by hpcsc** [comment]
@bpasero thanks for the review
these few days I'm struggling with these errors while merging from master:

screen shot 2017-01-23 at 4 06 42 pm
screen shot 2017-01-23 at 4 06 58 pm

do you happen to know what's the problem? I have no new commits since last time, just merge the changes from upstream. Already did `npm install`, `gulp clean` and `gulp compile` again but nothing works.

**Comment by bpasero** [comment]
@hpcsc what I typically do to get into a fresh state is: `git clean -xfd` and then `scripts/npm` install. Note that you should backup any work prior to calling `git clean`.

File: src/vs/base/parts/tree/browser/treeDefaults.ts [hunk 1]
```diff
@@ -91,8 +91,11 @@ export enum ClickBehavior {
 	ON_MOUSE_UP
 }
 
+export type WorkbenchOpenMode = 'singleClick' | 'doubleClick';
```
Comment (hpcsc) [inline src/vs/base/parts/tree/browser/treeDefaults.ts]: in that case, the configuration should be renamed as well? At the moment, it's named as `workbench.openMode`

**Comment by hpcsc** [comment]
@bpasero thanks, that indeed fixed the issue

Regarding using selection event, do you mean in controller we check for open mode and when it's suitable time to open editor, call:

```
tree.setSelection([element], payload);
```

to trigger selection event. Viewlets will subscribe to this selection event and does the opening of editor?

I think `tree.setSelection()` is still needed even when user does single click in double click open mode because that will highlight the current item in the tree. So we cannot use that to trigger open even. In other words, selection doesn't always mean open

File: src/vs/base/parts/tree/browser/treeDefaults.ts (hunk 1 above)
Comment (hpcsc) [inline src/vs/base/parts/tree/browser/treeDefaults.ts]: updated it to enum

File: src/vs/base/parts/tree/browser/treeDefaults.ts [hunk 2]
```diff
@@ -389,6 +395,16 @@ export class DefaultController implements _.IController {
...
 		return false;
 	}
+
+	protected setOpenMode(openMode: WorkbenchOpenMode) {
+		if (this.options.openMode !== openMode) {
+			this.options.openMode = openMode;
+		}
+	}
+
+	protected isInSingleClickOpenMode() {
```
Comment (hpcsc) [inline src/vs/base/parts/tree/browser/treeDefaults.ts]: renamed it to `openOnSingleClick`

File: src/vs/workbench/browser/treeController.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+interface IConfiguration {
+	workbench: {
+		openMode: string;
+	};
+}
+
+export class TreeControllerBase extends DefaultController {
+	constructor(
+		options: IControllerOptions,
```
Comment (hpcsc) [inline src/vs/workbench/browser/treeController.ts]: made it optional so child controllers don't need to pass in IControllerOptions if not needed

**Comment by bpasero** [comment]
@hpcsc sorry for not getting back earlier, just wanted to give a heads up that I started work on converting all trees to react on selection/focus event instead of doing the business of opening from within the controllers. We definitely want all trees to play by the same rules and open an item only when it has the selected state. I plan to continue this adoption in our next milestone, there are still some offenders.

As soon as this adoption is done, we can let the tree widget and list widget emit the selection event based on the new setting for mouse clicks (single click would focus and double click would select if the setting is enabled).

We still need to find a solution for the list widget because the list widget to my knowledge does not allow to control the behaviour on mouse click yet. 



PS: I am out on vacation next week.

**Comment by hpcsc** [comment]
@bpasero sorry for late reply, my work is a bit hectic recently
That's definitely a good news. However in that case most of the things in this pull request might be no longer relevant? Feel free to close this if it's no longer needed. If there's anything I can work on, I'm happy to help.

**Comment by bpasero** [comment]
Closing, this needs thinking time on my end.

**Comment by Jekton** [comment]
Is the Mac version supports this feature? It seems does not work in my Mac

**Pull request closed without merging**